transports/websocket-websys/src/lib.rs

```diff
@@ -0,0 +1,445 @@
+///
+/// See <https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/readyState>.
+#[derive(PartialEq)]
+enum ReadyState {
+    Connecting,
+    Open,
+    Closing,
+    Closed,
+}
+
+impl Connection {
+    fn new(socket: WebSocket) -> Self {
+        socket.set_binary_type(web_sys::BinaryType::Arraybuffer);
+
+        let open_waker = Rc::new(AtomicWaker::new());
+        let onopen_closure = Closure::<dyn FnMut(_)>::new({
+            let open_waker = open_waker.clone();
+            move |_| {
+                open_waker.wake();
+            }
+        });
+        socket.set_onopen(Some(onopen_closure.as_ref().unchecked_ref()));
+
+        let close_waker = Rc::new(AtomicWaker::new());
+        let onclose_closure = Closure::<dyn FnMut(_)>::new({
+            let close_waker = close_waker.clone();
+            move |_| {
+                close_waker.wake();
+            }
+        });
+        socket.set_onclose(Some(onclose_closure.as_ref().unchecked_ref()));
+
+        let errored = Rc::new(AtomicBool::new(false));
+        let onerror_closure = Closure::<dyn FnMut(_)>::new({
+            let errored = errored.clone();
+            move |_| {
+                errored.store(true, Ordering::SeqCst);
+            }
+        });
+        socket.set_onerror(Some(onerror_closure.as_ref().unchecked_ref()));
+
+        let read_buffer = Rc::new(Mutex::new(BytesMut::new()));
+        let new_data_waker = Rc::new(AtomicWaker::new());
+        let onmessage_closure = Closure::<dyn FnMut(_)>::new({
+            let read_buffer = read_buffer.clone();
+            let new_data_waker = new_data_waker.clone();
+            let errored = errored.clone();
+            move |e: MessageEvent| {
+                let data = js_sys::Uint8Array::new(&e.data());
+
+                let mut read_buffer = read_buffer.lock().unwrap();
+
+                if read_buffer.len() + data.length() as usize > MAX_BUFFER {
+                    log::warn!("Remote is overloading us with messages, closing connection");
+                    errored.store(true, Ordering::SeqCst);
+
+                    return;
+                }
+
+                read_buffer.extend_from_slice(&data.to_vec());
+                new_data_waker.wake();
+            }
+        });
+        socket.set_onmessage(Some(onmessage_closure.as_ref().unchecked_ref()));
+
+        let write_waker = Rc::new(AtomicWaker::new());
+        let on_buffered_amount_low_closure = Closure::<dyn FnMut(_)>::new({
+            let write_waker = write_waker.clone();
+            let socket = socket.clone();
+            move |_| {
+                if socket.buffered_amount() == 0 {
+                    write_waker.wake();
+                }
+            }
+        });
+        let buffered_amount_low_interval = window()
+            .expect("to have a window")
+            .set_interval_with_callback_and_timeout_and_arguments(
+                on_buffered_amount_low_closure.as_ref().unchecked_ref(),
+                500,
```

Comment:
These are milliseconds, correct? Isn't `500` rather large? On the rust-libp2p based IPFS bootstrap node I see a 50th percentile of 100 ms round-trip time. Thus with this interval, one is only able to send roughly one `MAX_BUFFER` per 5 round trips.

Would 10 milliseconds be too much work for a browser?

Reply:
> Would 10 milliseconds be too much work for a browser?

I have no idea to be honest. I can change it to a faster interval. At the same time, I don't think it matters much and I wouldn't invest time into "tuning" this transport. If somebody cares, they can open a PR and change this constant. In theory, we could probably also unset this interval while the transport is idle, i.e. once we fire the waker, we technically don't need to have this callback so we could unset it.

I really only want to land this in a minimal working state to remove JS code from the repository and make maintenance easier. I don't really care about the performance of this transport :)

Reply:
I suggest lowering it to `100ms` and adding a comment that it is chosen arbitrarily and likely worth tuning.